src/NLog.Targets.Syslog/AsyncLogger.cs

```diff
@@ -109,36 +112,20 @@ private Task ProcessQueueAsync(MessageBuilder messageBuilder, TaskCompletionSour
 
         private void Enqueue(AsyncLogEventInfo asyncLogEventInfo, int timeout)
         {
-            queue.TryAdd(asyncLogEventInfo, timeout, token);
+            if (!queue.TryAdd(asyncLogEventInfo, timeout, token))
+                asyncLogEventInfo.Continuation(new InvalidOperationException("AsyncLogger Queue Full"));
             if (InternalLogger.IsDebugEnabled)
                 InternalLogger.Debug("[Syslog] Enqueued '{0}'", asyncLogEventInfo.ToFormattedMessage());
         }
 
-        private AsyncLogEventInfo NewFlushCompletionMarker(TaskCompletionSource<object> tcs)
-        {
-            var asyncContinuation = new AsyncContinuation(_ => tcs.TrySetResult(null));
-            return new AsyncLogEventInfo(notLoggedLogEventInfo, asyncContinuation);
-        }
-
-        private void SignalFlushCompletionWhenIsMarker(AsyncLogEventInfo asyncLogEventInfo)
-        {
-            bool IsFlushCompletionMarker(AsyncLogEventInfo x) => x.LogEvent == notLoggedLogEventInfo;
-            void SignalFlushCompletion() => asyncLogEventInfo.Continuation(null);
-
-            if (IsFlushCompletionMarker(asyncLogEventInfo))
-            {
-                InternalLogger.Debug("[Syslog] AsyncLogger flushed");
-                SignalFlushCompletion();
-            }
-        }
-
         public void Dispose()
         {
-            cts.Cancel();
-            cts.Dispose();
-            queue.Dispose();
-            buffer.Dispose();
-            messageTransmitter.Dispose();
+            cts.Cancel();           // Signal cancel token
+            queue.CompleteAdding(); // Release all from queue, and block for adding new elements
```

Comment:
Not sure I understand. This is just me delaying dispose of token, until everyone is out of the burning house.

Reply:
As per [`CompleteAdding` docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1.completeadding):
"after a collection has been marked as complete for adding, adding to the collection is not permitted and attempts to remove from the collection will not wait when the collection is empty".
Flush not only can be called when terminating the client program, but also multiple times: add to the queue should be possible after a flush

Reply:
Dispose is called on Close, and not on Flush(). Hopefully a new BufferArray is allocated on Initialize.

Reply:
Moved into #201

Reply:
I think the order of calls can be improved, but `CompleteAdding` is not needed since [`cts.Cancel`](https://github.com/luigiberrettini/NLog.Targets.Syslog/blob/b9df517e502cef466eab35d9f0e156a1ca0bd43a/src/NLog.Targets.Syslog/AsyncLogger.cs#L137) has been called and the enqueuing [is based](https://github.com/luigiberrettini/NLog.Targets.Syslog/blob/b9df517e502cef466eab35d9f0e156a1ca0bd43a/src/NLog.Targets.Syslog/AsyncLogger.cs#L112) on the same token